vncviewer/MonitorArrangement.cxx

```diff
@@ -0,0 +1,414 @@
+    Fl::screen_xywh(x, y, w, h, i);
+    
+    top = y;
+    bottom = y + h;
+    left = x;
+    right = x + w;
+    
+    if (top < y_min) {
+      y_min = top;
+    }
+
+    if (bottom > y_max) {
+      y_max = bottom;
+    }
+
+    if (left < x_min) {
+      x_min = left;
+    }
+
+    if (right > x_max) {
+      x_max = right;
+    }
+  }
+  
+  return std::make_pair(x_max - x_min, y_max - y_min);
+}
+
+std::pair<int, int> MonitorArrangement::offset()
+{
+  double scale = this->scale();
+  std::pair<int, int> size = this->size();
+  std::pair<int, int> origin = this->origin();
+
+  int offset_x = (this->w()/2) - (size.first/2 * scale);
+  int offset_y = (this->h()/2) - (size.second/2 * scale);
+
+  return std::make_pair(offset_x + abs(origin.first), offset_y + abs(origin.second));
+}
+
+std::pair<int, int> MonitorArrangement::origin()
+{
+  int x, y, w, h, ox, oy;
+  ox = oy = 0;
+
+  for (int i = 0; i < Fl::screen_count(); i++) {
+    Fl::screen_xywh(x, y, w, h, i);
+
+    if (x < ox) {
+      ox = x;
+    }
+
+    if (y < oy) {
+      oy = y;
+    }
+  }
+
+  return std::make_pair(ox, oy);
+}
+
+bool MonitorArrangement::inside(
+  int top_y, int bottom_y, int left_x, int right_x, int x, int y)
+{
+  return (x > left_x) && (x < right_x) && (y > top_y) && (y < bottom_y);
+}
+
+std::string MonitorArrangement::description(int m)
+{
+  std::stringstream ss;
+  assert(m < (int) m_monitors.size());
+
+  // Get the platform specific name of the monitor. 
+  ss << name(m);
+
+  if (ss.str().empty()) {
+    int x, y, w, h;
+    Fl::screen_xywh(x, y, w, h, m);
+
+    // Fallback to showing resolution and position of monitor.
+    ss << w << "x" << h << "+" << x << "+" << y;
+  }
+
+  return ss.str();
+}
+
+std::string MonitorArrangement::name(int m)
+{
+  assert(m < (int) m_monitors.size());
+
+  int x, y, w, h;
+  std::stringstream ss;
+  Fl::screen_xywh(x, y, w, h, m);
+
+#if !defined(WIN32) && !defined(__APPLE__)
+#ifdef HAVE_XRANDR
+
+  int ev, err, xi_major;
+  fl_open_display();
+  assert(fl_display != NULL);
+
+  if (!XQueryExtension(fl_display, "RANDR", &xi_major, &ev, &err)) {
+    vlog.info("Unable to find X11 RANDR extension.");
+    return std::string();
+  }
+
+  XRRScreenResources *res = XRRGetScreenResources(fl_display, DefaultRootWindow(fl_display));
+  if (!res) {
+    vlog.error("Unable to get XRRScreenResources for fl_display.");
+    return std::string();
+  }
+
+  for (int i = 0; i < res->ncrtc; i++) {
+    XRRCrtcInfo *crtc = XRRGetCrtcInfo(fl_display, res, res->crtcs[i]);
+
+    if (!crtc) {
+      vlog.error("Unable to get XRRCrtcInfo for crtc %d.", i);
+      continue;
+    }
+
+    for (int j = 0; j < crtc->noutput; j++) {
+      bool monitor_found = (crtc->x == x) &&
+          (crtc->y == y) &&
+          (crtc->width == ((unsigned int) w)) &&
+          (crtc->height == ((unsigned int) h));
+
+      if (monitor_found) {
+        XRROutputInfo *output = XRRGetOutputInfo(fl_display, res, crtc->outputs[j]);
```

Comment:
Do we need to check monitors rather than outputs to support the latest RandR?

Reply:
Future improvement, added to the acceptance criterias for now.